File: src/main/java/com/github/unldenis/hologram/line/animated/StandardAnimatedLine.java
package com.github.unldenis.hologram.line.animated;

import com.github.unldenis.hologram.animation.Animation;
import com.github.unldenis.hologram.line.Line;
import java.util.Collection;
import java.util.Optional;
import org.bukkit.Bukkit;
import org.bukkit.entity.Player;
import org.bukkit.scheduler.BukkitTask;

public final class StandardAnimatedLine implements IAnimated {

  private final Line line;

  private Optional<Animation> animation = Optional.empty();
  private int taskID = -1;


  public StandardAnimatedLine(Line line) {
    this.line = line;
  }

  @Override
  public void setAnimation(Animation animation, Collection<Player> seeingPlayers) {
    this.animation = Optional.of(animation);

    Runnable taskR = () -> seeingPlayers.forEach(
        player -> animation.nextFrame(player, line));
    BukkitTask task;
    if (animation.async()) {
      task = Bukkit.getScheduler()
          .runTaskTimerAsynchronously(line.getPlugin(), taskR, animation.delay(),
              animation.delay());
    } else {
      task = Bukkit.getScheduler()
          .runTaskTimer(line.getPlugin(), taskR, animation.delay(), animation.delay());
    }
    this.taskID = task.getTaskId();
  }

  @Override
  public void removeAnimation() {
    if (taskID != -1) {
      Bukkit.getScheduler().cancelTask(taskID);
      taskID = -1;
    }
  }

  @Override
  public Optional<Animation> getAnimation() {
    return animation;
  }

  public int getTaskID() {
    return taskID;
  }
}

File: src/main/kotlin/com/github/unldenis/hologram/line/animated/StandardAnimatedLine.kt
package com.github.unldenis.hologram.line.animated

import com.github.unldenis.hologram.animation.Animation
import com.github.unldenis.hologram.line.Line
import org.bukkit.Bukkit
import org.bukkit.entity.Player
import java.util.*
import java.util.function.Consumer

class StandardAnimatedLine(private val line: Line) : IAnimated {

    override lateinit var animation: Animation
    private var taskID: Int = -1

    override fun setAnimation(animation: Animation, seeingPlayers: Collection<Player>) {
        this.animation = animation

        val taskR = Runnable {
            seeingPlayers.forEach(
                Consumer { player -> animation.nextFrame(player, line) })
        }
        val task = if (animation.async()) {
            Bukkit.getScheduler()
                .runTaskTimerAsynchronously(
                    line.plugin, taskR, animation.delay(),
                    animation.delay()
                )
        } else {
            Bukkit.getScheduler()
                .runTaskTimer(line.plugin, taskR, animation.delay(), animation.delay())
        }
        this.taskID = task.taskId
    }

    override fun removeAnimation() {
        if (taskID != -1) {
            Bukkit.getScheduler().cancelTask(taskID)
            taskID = -1
        }
    }
}
